std/mem.zig

```diff
@@ -170,72 +170,99 @@ pub fn dupe(allocator: &Allocator, comptime T: type, m: []const T) ![]T {
     return new_buf;
 }
 
+pub const Side = enum { LEFT = 1, RIGHT = 2, BOTH = 3, };
+
 /// Remove values from the beginning and end of a slice.
-pub fn trim(comptime T: type, slice: []const T, values_to_strip: []const T) []const T {
-    var begin: usize = 0;
+pub fn trim(comptime T: type, slice: []const T, values_to_strip: []const T, side: Side) []const T {
+     var begin: usize = 0;
     var end: usize = slice.len;
-    while (begin < end and indexOfScalar(T, values_to_strip, slice[begin]) != null) : (begin += 1) {}
-    while (end > begin and indexOfScalar(T, values_to_strip, slice[end - 1]) != null) : (end -= 1) {}
+
+    // Replace later with bitwise, Zig seems to require cast for enum bitwise?
+    if (side == Side.LEFT or side == Side.BOTH) {
+        while (begin < end and indexOfScalar(T, values_to_strip, slice[begin]) != null) : (begin += 1) {}
+    }
+
+    if (side == Side.RIGHT or side == Side.BOTH) {
+        while (end > begin and indexOfScalar(T, values_to_strip, slice[end - 1]) != null) : (end -= 1) {}
+    }
+
     return slice[begin..end];
 }
 
 test "mem.trim" {
-    assert(eql(u8, trim(u8, " foo\n ", " \n"), "foo"));
-    assert(eql(u8, trim(u8, "foo", " \n"), "foo"));
+    assert(eql(u8, trim(u8, " foo\n ", " \n", Side.BOTH), "foo"));
+    assert(eql(u8, trim(u8, "foo", " \n", Side.BOTH), "foo"));
+    assert(eql(u8, trim(u8, " foo ", " ", Side.LEFT), "foo "));
 }
 
 /// Linear search for the index of a scalar value inside a slice.
 pub fn indexOfScalar(comptime T: type, slice: []const T, value: T) ?usize {
-    return indexOfScalarPos(T, slice, 0, value);
+    return indexOfScalarPos(T, slice, 0, value, false);
 }
 
-pub fn indexOfScalarPos(comptime T: type, slice: []const T, start_index: usize, value: T) ?usize {
+pub fn indexOfScalarPos(comptime T: type, slice: []const T, start_index: usize, value: T, highest: bool) ?usize {
     var i: usize = start_index;
+    var out: ?usize = null;
+
     while (i < slice.len) : (i += 1) {
-        if (slice[i] == value)
-            return i;
+        if (slice[i] == value) {
+            out = i;
+            if (!highest) break;
+        }
     }
-    return null;
+
+    return out;
 }
 
 pub fn indexOfAny(comptime T: type, slice: []const T, values: []const T) ?usize {
-    return indexOfAnyPos(T, slice, 0, values);
+    return indexOfAnyPos(T, slice, 0, values, false);
 }
 
-pub fn indexOfAnyPos(comptime T: type, slice: []const T, start_index: usize, values: []const T) ?usize {
+pub fn indexOfAnyPos(comptime T: type, slice: []const T, start_index: usize, values: []const T, highest: bool) ?usize {
     var i: usize = start_index;
+    var out: ?usize = null;
+
     while (i < slice.len) : (i += 1) {
         for (values) |value| {
-            if (slice[i] == value)
-                return i;
+            if (slice[i] == value) {
+                out = i;
+                if (!highest) break;
```

Comment:
`lastIndexOfAnyPos` should have a better algorithm than this. at worst searching from the end, at best reverse boyer-moore or something like that.